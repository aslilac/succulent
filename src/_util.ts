type IsAny<T> = unknown extends T ? (T extends {} ? T : never) : never;
type NotAny<T> = T extends IsAny<T> ? never : T;

export function assertType<T, X extends T>(x: NotAny<X>) {}

/**
 * This should always be an error (hence the ts-expect-error), and
 * is put here to make sure this is always true whenever assertType
 * is imported. Because assertType is a noop, the type error isn't
 * actually dangerous.
 */
// @ts-expect-error
assertType<unknown, any>({} as any);
