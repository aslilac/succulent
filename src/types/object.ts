import { Schema, SchemaBase } from "../schema";

export const $anyobject = new Schema(
	(x: unknown): x is object => typeof x === "object" && x != null,
);

export function $object<T extends object>(template: {
	[K in keyof T]: SchemaBase<T[K]>;
}): Schema<T> {
	return new Schema(
		(x: unknown): x is T =>
			typeof x === "object" &&
			x != null &&
			// [
			// 	...Object.getOwnPropertyNames(template),
			// 	...Object.getOwnPropertySymbols(template),
			// ]
			Reflect.ownKeys(template).every((key) =>
				// @ts-expect-error - Can't quite get these types
				Schema.check(template[key], x[key]),
			),
	);
}
