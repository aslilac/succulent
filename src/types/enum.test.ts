/// <reference types="jest" />
import { assertType } from "../_util";
import { check, is, $enum } from "../index";
import { $object } from "./object";

enum Fruit {
	Apple,
	Banana,
	Mango,
	Peach,
	Strawberry,
}

enum FunnyNumber {
	SixtyNine = 69,
	FourTwenty = 420,
}

enum Friend {
	August = "August",
	Bandit = "Bandit",
	Dot = "Dot",
	Mady = "Mady",
	Toby = "Toby",
}

enum Message {
	Hello = "Hello, friend! How are you doing?",
	Bye = "Goodbye, friend! I'll see you later!",
}

test("$enum", () => {
	const $Fruit = $enum(Fruit);
	expect(is(Fruit.Apple, $Fruit)).toBe(true);
	expect(is(0, $Fruit)).toBe(true);
	expect(is("Apple", $Fruit)).toBe(false); // because `"Apple" !== Fruit.Apple`
	expect(is(5, $Fruit)).toBe(false); // because the enum ends at Strawberry = 4
	expect($Fruit.displayName).toBe("enum { Apple, Banana, Mango, Peach, Strawberry }");

	const $FunnyNumber = $enum(FunnyNumber);
	expect(is(FunnyNumber.SixtyNine, $FunnyNumber)).toBe(true);
	expect(is(420, $FunnyNumber)).toBe(true);
	expect(is("FourTwenty", $FunnyNumber)).toBe(false); // because `"FourTwenty" !== FunnyNumber.FourTwenty`
	expect(is(421, $FunnyNumber)).toBe(false);
	expect($FunnyNumber.displayName).toBe("enum { SixtyNine, FourTwenty }");

	const $Friend = $enum(Friend);
	expect(is(Friend.August, $Friend)).toBe(true);
	expect(is("Bandit", $Friend)).toBe(true); // because `"Bandit" === Friend.Bandit`
	expect($Friend.displayName).toBe("enum { August, Bandit, Dot, Mady, Toby }");

	const $Message = $enum(Message, { displayName: "Message" });
	expect(is(Message.Hello, $Message)).toBe(true);
	expect(is("Bye", $Message)).toBe(false); // because `"Bye" === Message.Bye`

	const $MessageToFriend = $object({
		friend: $Friend,
		message: $Message,
	});

	const good = { friend: Friend.Bandit, message: Message.Hello };
	const bad = { friend: "Tony", message: null };
	expect(is(good, $MessageToFriend)).toBe(true);
	expect(() => check(bad, $MessageToFriend)).toThrowErrorMatchingSnapshot();

	function _(x: unknown) {
		if (is(x, $Fruit)) assertType<Fruit, typeof x>(x);
		if (is(x, $FunnyNumber)) assertType<FunnyNumber, typeof x>(x);
		if (is(x, $Friend)) assertType<Friend, typeof x>(x);
		if (is(x, $Message)) assertType<Message, typeof x>(x);
	}
});
