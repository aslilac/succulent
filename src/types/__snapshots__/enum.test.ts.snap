// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`$enum 1`] = `
"Expected {friend: enum { August, Bandit, Dot, Mady, Toby }, message: Message}, got [object Object]
  Expected property friend to have type enum { August, Bandit, Dot, Mady, Toby }
    Expected enum { August, Bandit, Dot, Mady, Toby }, got \\"Tony\\"
  Expected property message to have type Message
    Expected Message, got null"
`;
